src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -3,17 +3,129 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
 
 namespace System.Dynamic.Utils
 {
     internal static partial class CollectionExtensions
     {
+        public static T[] AddFirst<T>(this IList<T> list, T item)
+        {
+            T[] res = new T[list.Count + 1];
+            res[0] = item;
+            list.CopyTo(res, 1);
+            return res;
+        }
+
         public static T[] AddLast<T>(this IList<T> list, T item)
```

Comment:
Noticed that this only operates on arrays, so we can change the type from `IList<T>` to `T[]` at some point. Will do that in a separate PR.